src/ray/common/test/ray_syncer_test.cc

```diff
@@ -306,14 +306,10 @@ struct SyncerServerTest {
     auto start = steady_clock::now();
 
     while (duration_cast<seconds>(steady_clock::now() - start).count() <= time_s) {
-      std::promise<bool> p;
-      auto f = p.get_future();
-      io_context.post([&p, predicate]() mutable { p.set_value(predicate()); }, "TEST");
-      if (f.get()) {
```

Comment:
`std::future<T>::get` is a blocking call, I don't see any meaning to use future / promise / io_context here.

Reply:
I think the point is that it wants the predicate() func to be run in the ioctx thread, while also wants the post to be sync (blocking). If you look at callers of WaitUntil, many of them access ray_syncer_ which may not be thread safe.

Reply:
OK I see, we need to queue the functor to make it executed sequentially.